### PR TITLE
[Lottie Exporter] Export outlines without variable widths

### PR DESCRIPTION
**Pull request by saurabh-ramola**
Added support for exporting outlines without variable widths. A few tweaks needed but there is a huge decrease in file size on checking so far :smiley: .
@AnishGG please check

### CODE REVIEW

**Comment by morevnaproject** [comment]
Question: how user can choose to export with variable outline width or not?

**Comment by AnishGG** [comment]
> Question: how user can choose to export with variable outline width or not?

For now, @saurabh-ramola used "Exporting without variable width's" as the default option. But I think what we can do is: User selects Menu -> Export -> show a dialogue in which user selects whether to export with/without variable widths. 
Finally we can call the lottie-exporter with a flag based on the option selected by the user. 

@morevnaproject your views here please. ^__^
@saurabh-ramola in the mean time, you can look at the flag option and implement it if possible.

**Comment by saurabh-ramola** [comment]
> > Question: how user can choose to export with variable outline width or not?
> 
> For now, @saurabh-ramola used "Exporting without variable width's" as the default option. But I think what we can do is: User selects Menu -> Export -> show a dialogue in which user selects whether to export with/without variable widths.
> Finally we can call the lottie-exporter with a flag based on the option selected by the user.
> 
> @morevnaproject your views here please. ^__^
> @saurabh-ramola in the mean time, you can look at the flag option and implement it if possible.

@AnishGG I have added an argument `-without_variable_width` which is False by default but user has the option to set it to True which will export outlines without variable widths when it calls the python file. But I don't know how to implement the GUI for it and how to call it from synfig-studio.

**Comment by morevnaproject** [comment]
Question: how user can choose to export with variable outline width or not?

> I have added an argument `-without_variable_width` which is False by default but user has the option to set it to True which will export outlines without variable widths when it calls the python file. But I don't know how to implement the GUI for it and how to call it from synfig-studio.

I see two possibilities for this:

**1. Add more export options in `plugin.xml` file.**

Our plugin system now got improvements from @mbasaglia which allow to define multiple plugins in `plugin.xml`  file. It is possible to call same python file, but specify different arguments. See specification here - https://github.com/synfig/synfig/issues/1309#issuecomment-606829467

Currently, `plugin.xml` file of Lottie exporter has following entries:
```
<exporter>
       <extension>json</extension>
       <description>Lottie Animation (*.json)</description>
       <exec stdout="log">lottie-exporter.py</exec>
   </exporter>
   <exporter>
       <extension>html</extension>
       <description>Lottie HTML Preview (*.html)</description>
       <exec stdout="log">lottie-exporter.py</exec>
   </exporter>
```
you can add two more:
* Lottie Animation [simplified outlines] (*.json)
* Lottie HTML Preview [simplified outlines] (*.html)

**2. Other way is to add parameter for Group Layer, called "Fixed outline width" [bool].**

This parameter will tell Synfig to render all outline layers inside with fixed line width.

This approach is more complex than previous, but it has several advantages:
* It is possible to define which parts of animation will use ordinary outline width and which will use fixed outline width (optimized outlines)
* User can see the result directly in Synfig before exporting.

Regarding implementation, I recommend to look at how "Outline Grow" parameter works - https://wiki.synfig.org/Group_Layer#Outline_Grow_Parameter
Its principle is the same - it modifies how outlines are rendered inside of group.

**Comment by saurabh-ramola** [comment]
> Question: how user can choose to export with variable outline width or not?
> 
> > I have added an argument `-without_variable_width` which is False by default but user has the option to set it to True which will export outlines without variable widths when it calls the python file. But I don't know how to implement the GUI for it and how to call it from synfig-studio.
> 
> I see two possibilities for this:
> 
> **1. Add more export options in `plugin.xml` file.**
> 
> Our plugin system now got improvements from @mbasaglia which allow to define multiple plugins in `plugin.xml` file. It is possible to call same python file, but specify different arguments. See specification here - [#1309 (comment)](https://github.com/synfig/synfig/issues/1309#issuecomment-606829467)
> 
> Currently, `plugin.xml` file of Lottie exporter has following entries:
> 
> ```
> <exporter>
>        <extension>json</extension>
>        <description>Lottie Animation (*.json)</description>
>        <exec stdout="log">lottie-exporter.py</exec>
>    </exporter>
>    <exporter>
>        <extension>html</extension>
>        <description>Lottie HTML Preview (*.html)</description>
>        <exec stdout="log">lottie-exporter.py</exec>
>    </exporter>
> ```
> 
> you can add two more:
> 
> * Lottie Animation [simplified outlines] (*.json)
> * Lottie HTML Preview [simplified outlines] (*.html)
> 
> **2. Other way is to add parameter for Group Layer, called "Fixed outline width" [bool].**
> 
> This parameter will tell Synfig to render all outline layers inside with fixed line width.
> 
> This approach is more complex than previous, but it has several advantages:
> 
> * It is possible to define which parts of animation will use ordinary outline width and which will use fixed outline width (optimized outlines)
> * User can see the result directly in Synfig before exporting.
> 
> Regarding implementation, I recommend to look at how "Outline Grow" parameter works - https://wiki.synfig.org/Group_Layer#Outline_Grow_Parameter
> Its principle is the same - it modifies how outlines are rendered inside of group.

Awesome, I'll start looking into this :)

**Comment by mbasaglia** [comment]
I think the best option would be to extend the plugin system further, and allow options to be specified in the plugin.xml that would show a dialog before running the script.

Eg:
```
<exporter>
...
<option label="Simplify outlines" type="bool"/>

</exporter>
```

Otherwise you could display such a dialog from python

**Comment by saurabh-ramola** [comment]
@ice0 @AYESDIE , I created a new file to make outlines without variable widths and added the new file name to the CMakeLists.txt file but when I try to export an outline it gives me this error. I initially thought this error was due to the new file not compiling but even on recompiling it doesn't export. Would you be knowing the reason behind this error because on exporting the file from the command line(executing the python file), it runs without any problem.
image

**Comment by saurabh-ramola** [comment]
@mbasaglia , I added the option `--without_variable_width` which tells the python program to export outlines without variable width when set to True. I added the argument like this https://github.com/synfig/synfig/blob/8ea969822419ade47a498f24e6d7061fa54736aa/synfig-studio/plugins/lottie-exporter/lottie-exporter.py#L136 and made changes into the plugin file like this https://github.com/synfig/synfig/blob/8ea969822419ade47a498f24e6d7061fa54736aa/synfig-studio/plugins/lottie-exporter/plugin.xml.in#L14-L21 .

However, after compiling the changes,I exported the file using the new options I had added but the export did not pass the `--without_variable_width` argument and I still get the output as if the argument was set to False. Could you explain to me where I went wrong.

TIA

**Comment by mbasaglia** [comment]
Options aren't implemented yet, but you can call a different python script

**Comment by saurabh-ramola** [comment]
Added another script `export_without_variable_width` in the `plugin.xml` file to be called when user selects the option of exporting without variable widths.

File: synfig-studio/plugins/lottie-exporter/plugin.xml.in [hunk 1]
```diff
@@ -11,4 +11,14 @@
...
        <exec stdout="log">lottie-exporter.py</exec>
    </exporter>
+   <exporter>
+       <extension>html</extension>
+       <_description>Lottie HTML Preview without variable widths (*.html)</_description>
+       <exec stdout="log" interpreter="python">export_without_variable_width.py</exec>
+   </exporter>
+   <exporter>
+       <extension>json</extension>
+       <_description>Lottie HTML Preview without variable widths(*.json)</_description>
```
Comment (morevnaproject) [inline synfig-studio/plugins/lottie-exporter/plugin.xml.in:21]: -> `<_description>Lottie Animation without variable widths (*.json)</_description>`

**Review by morevnaproject — CHANGES_REQUESTED** [review]

**Comment by morevnaproject** [comment]
I have made attempt to export a simple animated outlines and for some reason it doesn't exports animation.

File - [lottie-outlines.sifz.zip](https://github.com/synfig/synfig/files/5130922/lottie-outlines.sifz.zip)

Also, I have noticed the resulting outline is much thicker (if we compare with same file exported with variable width) -
screenshot_001

File: synfig-studio/plugins/lottie-exporter/plugin.xml.in (hunk 1 above)
Comment (rodolforg) [reply synfig-studio/plugins/lottie-exporter/plugin.xml.in:21]: Why does the tag name start with `_`?

Comment (morevnaproject) [reply synfig-studio/plugins/lottie-exporter/plugin.xml.in:21]: This indicates tag contents as translatable.

**Comment by saurabh-ramola** [comment]
Found the issue, there's a mistake in naming the outline dictionaries, fixing it now.

File: synfig-studio/plugins/lottie-exporter/plugin.xml.in (hunk 1 above)
Comment (saurabh-ramola) [reply synfig-studio/plugins/lottie-exporter/plugin.xml.in]: @morevnaproject  changed the description content.

**Comment by saurabh-ramola** [comment]
@morevnaproject can you check now?

**Comment by saurabh-ramola** [comment]
> Found the issue, there's a mistake in naming the outline dictionaries, fixing it now.

This wasn't the issue, the issue was a faulty frame appearing at the end of the dictionary.

**Comment by saurabh-ramola** [comment]
The failing build happens as some of the PRs haven't been merged yet, but have passed review.

**Comment by morevnaproject** [comment]
> The failing build happens as some of the PRs haven't been merged yet, but have passed review.

I have fixed this now. ^__^

**Comment by morevnaproject** [comment]
> @morevnaproject can you check now?

Works good for me now! ^__^

**Comment by ice0** [comment]
Merged. Thank you!

**Pull request merged**